### ISSUE TITLE
Compilation errors on master

### ISSUE THREAD

**Issue by vadimkantorov** [facebookresearch/wav2letter#301]
I run the Dockerfile-CUDA and get the following errors:

```
/root/wav2letter/Decode.cpp: In lambda function:
/root/wav2letter/Decode.cpp:319:19: error: 'make_unique' is not a member of 'std'
         decoder = std::make_unique<WordLMDecoder>(
                   ^
/root/wav2letter/Decode.cpp:319:19: note: suggested alternative:
In file included from /usr/local/include/flashlight/common/Serialization-inl.h:17:0,
                 from /usr/local/include/flashlight/common/Serialization.h:195,
                 from /usr/local/include/flashlight/autograd/Variable.h:25,
                 from /usr/local/include/flashlight/autograd/Utils.h:11,
                 from /usr/local/include/flashlight/autograd/autograd.h:20,
                 from /usr/local/include/flashlight/flashlight.h:11,
                 from /root/wav2letter/Decode.cpp:17:
/usr/local/include/flashlight/common/CppBackports.h:104:38: note:   'fl::cpp::make_unique'
 typename _unique_if<T>::_known_bound make_unique(Args&&...) = delete;
                                      ^
/root/wav2letter/Decode.cpp:319:49: error: expected primary-expression before '>' token
         decoder = std::make_unique<WordLMDecoder>(
                                                 ^
/root/wav2letter/Decode.cpp:330:21: error: 'make_unique' is not a member of 'std'
           decoder = std::make_unique<TokenLMDecoder>(
                     ^

/root/wav2letter/Decode.cpp:330:21: note: suggested alternative:
In file included from /usr/local/include/flashlight/common/Serialization-inl.h:17:0,
                 from /usr/local/include/flashlight/common/Serialization.h:195,
                 from /usr/local/include/flashlight/autograd/Variable.h:25,
                 from /usr/local/include/flashlight/autograd/Utils.h:11,
                 from /usr/local/include/flashlight/autograd/autograd.h:20,
                 from /usr/local/include/flashlight/flashlight.h:11,
                 from /root/wav2letter/Decode.cpp:17:
/usr/local/include/flashlight/common/CppBackports.h:104:38: note:   'fl::cpp::make_unique'
 typename _unique_if<T>::_known_bound make_unique(Args&&...) = delete;
                                      ^
/root/wav2letter/Decode.cpp:330:52: error: expected primary-expression before '>' token
           decoder = std::make_unique<TokenLMDecoder>(
                                                    ^
/root/wav2letter/Decode.cpp:342:21: error: 'make_unique' is not a member of 'std'
           decoder = std::make_unique<LexiconFreeDecoder>(
                     ^
/root/wav2letter/Decode.cpp:342:21: note: suggested alternative:
In file included from /usr/local/include/flashlight/common/Serialization-inl.h:17:0,
                 from /usr/local/include/flashlight/common/Serialization.h:195,
                 from /usr/local/include/flashlight/autograd/Variable.h:25,
                 from /usr/local/include/flashlight/autograd/Utils.h:11,
                 from /usr/local/include/flashlight/autograd/autograd.h:20,
                 from /usr/local/include/flashlight/flashlight.h:11,
                 from /root/wav2letter/Decode.cpp:17:
/usr/local/include/flashlight/common/CppBackports.h:104:38: note:   'fl::cpp::make_unique'
 typename _unique_if<T>::_known_bound make_unique(Args&&...) = delete;
                                      ^
/root/wav2letter/Decode.cpp:342:56: error: expected primary-expression before '>' token
           decoder = std::make_unique<LexiconFreeDecoder>(


                                                        ^
CMakeFiles/DitherTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::TrieNode const*, w2l::LexiconD
ecoderState const*, float, int, w2l::TrieLabel const*, bool)':
LexiconDecoder.cpp:(.text+0x3e9): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/DitherTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::decodeEnd()':
LexiconDecoder.cpp:(.text+0x166f): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/DitherTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::LexiconFreeDecoderStat
e const*, float, int, bool)':
LexiconFreeDecoder.cpp:(.text+0xb8d): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/DitherTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::decodeStep(float const*, int, int)':
LexiconFreeDecoder.cpp:(.text+0x1a3d): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
LexiconFreeDecoder.cpp:(.text+0x1c34): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/DitherTest.dir/__/decoder/LexiconFreeDecoder.cpp.o:LexiconFreeDecoder.cpp:(.text+0x27db): more undefined references to `w2l::isGoodCandidate(float&, float, float)' follo
w
collect2: error: ld returned 1 exit status
src/tests/CMakeFiles/DitherTest.dir/build.make:1427: recipe for target 'src/tests/DitherTest' failed
make[2]: *** [src/tests/DitherTest] Error 1
CMakeFiles/Makefile2:361: recipe for target 'src/tests/CMakeFiles/DitherTest.dir/all' failed
make[1]: *** [src/tests/CMakeFiles/DitherTest.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....


make[1]: *** Waiting for unfinished jobs....
[  8%] Building CXX object CMakeFiles/Decoder.dir/src/common/Utils.cpp.o
CMakeFiles/CeplifterTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::TrieNode const*, w2l::Lexic
onDecoderState const*, float, int, w2l::TrieLabel const*, bool)':
LexiconDecoder.cpp:(.text+0x3e9): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/CeplifterTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::decodeEnd()':
LexiconDecoder.cpp:(.text+0x166f): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/CeplifterTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::LexiconFreeDecoderS
tate const*, float, int, bool)':
LexiconFreeDecoder.cpp:(.text+0xb8d): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/CeplifterTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::decodeStep(float const*, int, int)':
LexiconFreeDecoder.cpp:(.text+0x1a3d): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
LexiconFreeDecoder.cpp:(.text+0x1c34): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/CeplifterTest.dir/__/decoder/LexiconFreeDecoder.cpp.o:LexiconFreeDecoder.cpp:(.text+0x27db): more undefined references to `w2l::isGoodCandidate(float&, float, float)' fo
llow
collect2: error: ld returned 1 exit status
src/tests/CMakeFiles/CeplifterTest.dir/build.make:1427: recipe for target 'src/tests/CeplifterTest' failed
make[2]: *** [src/tests/CeplifterTest] Error 1
CMakeFiles/Makefile2:321: recipe for target 'src/tests/CMakeFiles/CeplifterTest.dir/all' failed
make[1]: *** [src/tests/CMakeFiles/CeplifterTest.dir/all] Error 2
[  8%] Building CXX object CMakeFiles/Decoder.dir/src/criterion/ConnectionistTemporalClassificationCriterion.cpp.o
CMakeFiles/W2lModuleTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::TrieNode const*, w2l::Lexic
onDecoderState const*, float, int, w2l::TrieLabel const*, bool)':
LexiconDecoder.cpp:(.text+0x3e9): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/W2lModuleTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::decodeEnd()':
LexiconDecoder.cpp:(.text+0x166f): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/W2lModuleTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::LexiconFreeDecoderS
tate const*, float, int, bool)':


LexiconFreeDecoder.cpp:(.text+0xb8d): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/W2lModuleTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::decodeStep(float const*, int, int)':
LexiconFreeDecoder.cpp:(.text+0x1a3d): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
LexiconFreeDecoder.cpp:(.text+0x1c34): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/W2lModuleTest.dir/__/decoder/LexiconFreeDecoder.cpp.o:LexiconFreeDecoder.cpp:(.text+0x27db): more undefined references to `w2l::isGoodCandidate(float&, float, float)' fo
llow
collect2: error: ld returned 1 exit status
src/tests/CMakeFiles/W2lModuleTest.dir/build.make:1427: recipe for target 'src/tests/W2lModuleTest' failed
make[2]: *** [src/tests/W2lModuleTest] Error 1
CMakeFiles/Makefile2:441: recipe for target 'src/tests/CMakeFiles/W2lModuleTest.dir/all' failed
make[1]: *** [src/tests/CMakeFiles/W2lModuleTest.dir/all] Error 2
[  8%] Building CXX object CMakeFiles/Decoder.dir/src/criterion/CriterionUtils.cpp.o
CMakeFiles/DctTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::TrieNode const*, w2l::LexiconDeco
derState const*, float, int, w2l::TrieLabel const*, bool)':
LexiconDecoder.cpp:(.text+0x3e9): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/DctTest.dir/__/decoder/LexiconDecoder.cpp.o: In function `w2l::LexiconDecoder::decodeEnd()':
LexiconDecoder.cpp:(.text+0x166f): undefined reference to `w2l::isGoodCandidate(float&, float, float)'
CMakeFiles/DctTest.dir/__/decoder/LexiconFreeDecoder.cpp.o: In function `w2l::LexiconFreeDecoder::candidatesAdd(std::shared_ptr<w2l::LMState> const&, w2l::LexiconFreeDecoderState c
onst*, float, int, bool)':

# and more similar errors
```

(I comment out CharLMDecoder.cpp from CMakeLists.txt to get around https://github.com/facebookresearch/wav2letter/issues/299 )

**Comment by vadimkantorov** [comment]
These errors start from 0b504c199c3c49704703f7b456f745010b298bdc

**Comment by vadimkantorov** [comment]
Maybe a newer version of gcc compiler should be selected in the Dockerfile, currently a quite old 5.4.0 is used

**Comment by tlikhomanenko** [comment]
@vadimkantorov  

Thanks for reporting. Will provide the correct Dockerfiles in few days.

**Comment by vadimkantorov** [comment]
@tlikhomanenko While on this, could you please advise if I need to use nvidia-docker, or just docker should do?

**Comment by jacobkahn** [comment]
@vadimkantorov — you'll need to use `nvidia-docker`

**Comment by jacobkahn** [comment]
Fixed in b9fa15c46a487db6e4afce6c83d30ef76165a48d. We're sticking with gcc 4.8.5.
